### PR TITLE
[BugFix] fix memory usage issue of spillable hash join

### PR DESCRIPTION
**Pull request by silverbullet233**
fix two memory usage issues of spillable hash join operator: 

1. if hash join build operator change to spill mode just before set_finishing phase, current implementation didn't handle it correctly and still execute as non-spill mode, here I fix it
2. in hash join probe side, it will load all in-memory partitions first, but in build side, it doesn't consider the total partition bytes during flush, this may lead to uncontrollable memory usage when the probe loads the partition. here I added a strategy to ensure that the partition resident in memory is not too large after the final flush.

after applying this change, tpc-ds Q29 can run successfully under 1T scale with 20G query_mem_limit in one 16c BE instance.

## What type of PR is this:
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Review by stdpain — DISMISSED** [review]

File: be/src/exec/pipeline/hashjoin/spillable_hash_join_build_operator.cpp [hunk 1]
```diff
@@ -70,10 +70,24 @@ bool SpillableHashJoinBuildOperator::need_input() const {
...
 
-    if (!_join_builder->spiller()->spilled()) {
+    if (spill_strategy() == spill::SpillStrategy::NO_SPILL ||
+        (!_join_builder->spiller()->spilled() &&
+         _join_builder->hash_join_builder()->hash_table().get_row_count() == 0)) {
         return HashJoinBuildOperator::set_finishing(state);
     }
 
+    // if this operator is changed to spill mode just before set_finishing,
+    // we should create spill task
```
Comment (fzhedu) [inline be/src/exec/pipeline/hashjoin/spillable_hash_join_build_operator.cpp:80]: CHECK FORCE_SPILLED  before actually to spill?

File: be/src/exec/pipeline/hashjoin/spillable_hash_join_build_operator.cpp [hunk 2]
```diff
@@ -70,10 +70,24 @@ bool SpillableHashJoinBuildOperator::need_input() const {
...
+    if (!_join_builder->spiller()->spilled()) {
+        DCHECK(_is_first_time_spill);
+        _is_first_time_spill = false;
+        auto& ht = _join_builder->hash_join_builder()->hash_table();
+        RETURN_IF_ERROR(init_spiller_partitions(state, ht));
+
+        _hash_table_slice_iterator = _convert_hash_map_to_chunk();
+        RETURN_IF_ERROR(_join_builder->append_spill_task(state, _hash_table_slice_iterator));
```
Comment (fzhedu) [inline be/src/exec/pipeline/hashjoin/spillable_hash_join_build_operator.cpp:88]: where can we known this change of spill mode? from profile or log?

Comment (silverbullet233) [reply be/src/exec/pipeline/hashjoin/spillable_hash_join_build_operator.cpp:88]: > where can we known this change of spill mode? from profile or log?

we can use SpillerMetrics in profile to check wether the spill mode changed, but the first spill  is triggered when set_finishing or push_chunk, the user does not need to perceive it, and it cannot be seen through the profile. 
I add some logs to observe it during my test.

File: be/src/exec/pipeline/hashjoin/spillable_hash_join_build_operator.cpp (hunk 1 above)
Comment (silverbullet233) [reply be/src/exec/pipeline/hashjoin/spillable_hash_join_build_operator.cpp:81]: > CHECK FORCE_SPILLED before actually to spill?

add a DCHECK here

**Comment by silverbullet233** [comment]
> ```
> after applying this change, tpc-ds Q29 can run successfully under 1T scale with 20G query_mem_limit in one 16c BE instance.
> ```
> 
> any reports about testing spill in 1T tpcds? which costs too much memory and which part takes the 20GB available memory? can we further reduce it?

Yes, some data need to be sorted out, I will send a test report later

**Review by fzhedu — APPROVED** [review]
LGTM

**Review by stdpain — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Pull request merged**